File: src/1438.py
import logging
import sys
from bisect import bisect_left, bisect_right
from collections import Counter, defaultdict, deque
from functools import cache, lru_cache
from heapq import heapify, heappop, heappush
from itertools import accumulate
from typing import List, Optional


def logging_setting():
    LOG_LEVEL = logging.INFO    # DEBUG, INFO, WARNING, ERROR, CRITICAL
    if __debug__:
        LOG_LEVEL = logging.DEBUG

    logging.basicConfig(
        stream=sys.stderr,
        level=LOG_LEVEL,
        format="%(levelname)-6s %(asctime)s %(module)s:%(funcName)s:%(lineno)-4s %(message)s",
        datefmt='%Y/%m/%d %H:%M:%S')


class Solution:
    def longestSubarray(self, nums: List[int], limit: int) -> int:
        retVal = 0

        maxDeque = deque()
        minDeque = deque()
        numsSize = len(nums)
        left = 0
        right = 0
        for right in range(numsSize):
            # Maintain the maxDeque in decreasing order
            while (maxDeque) and (maxDeque[-1] < nums[right]):
                maxDeque.pop()
            maxDeque.append(nums[right])

            # Maintain the minDeque in increasing order
            while (minDeque) and (minDeque[-1] > nums[right]):
                minDeque.pop()
            minDeque.append(nums[right])

            # Check if the current window exceeds the limit
            while maxDeque[0] - minDeque[0] > limit:
                # Remove the elements that are out of the current window
                if maxDeque[0] == nums[left]:
                    maxDeque.popleft()
                if minDeque[0] == nums[left]:
                    minDeque.popleft()

                left += 1

            retVal = max(retVal, right - left + 1)

        return retVal


if __name__ == "__main__":
    logging_setting()

    try:
        logging.info("sys.version: %s", sys.version)
        print()

        pSolution = Solution()
        for nums, limit in zip([[8, 2, 4, 7], [10, 1, 2, 4, 7, 2], [4, 2, 2, 2, 4, 4, 2, 2]], [4, 5, 0]):
            # /* Example
            #  *  Input: nums = [8,2,4,7], limit = 4
            #  *  Output: 2
            #  *
            #  *  Input: nums = [10,1,2,4,7,2], limit = 5
            #  *  Output: 4
            #  *
            #  *  Input: nums = [4,2,2,2,4,4,2,2], limit = 0
            #  *  Output: 3
            #  */
            logging.info("Input: nums = %s, limit = %s", nums, limit)

            retVal = pSolution.longestSubarray(nums, limit)
            logging.info("Output: %s", retVal)

            print()
    except KeyboardInterrupt as exception:
        logging.error("%s: %s", exception.__class__.__name__,
                      exception, exc_info=True)
        pass
    except Exception as exception:
        logging.error("%s: %s", exception.__class__.__name__,
                      exception, exc_info=True)
        pass

    sys.exit(0)

File: src/1438.c
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int longestSubarray(int* nums, int numsSize, int limit) {
    int retVal = 0;

    int maxDequeLeft = 0;
    int maxDequeRight = 0;
    int maxDeque[numsSize];
    memset(maxDeque, 0, sizeof(maxDeque));

    int minDequeLeft = 0;
    int minDequeRight = 0;
    int minDeque[numsSize];
    memset(minDeque, 0, sizeof(minDeque));

    int left = 0;
    int right = 0;
    for (right = 0; right < numsSize; right++) {
        // Maintain the maxDeque in decreasing order
        while ((maxDequeLeft < maxDequeRight) && (maxDeque[maxDequeRight - 1] < nums[right])) {
            maxDequeRight--;
        }
        maxDeque[maxDequeRight++] = nums[right];

        // Maintain the minDeque in increasing order
        while ((minDequeLeft < minDequeRight) && (minDeque[minDequeRight - 1] > nums[right])) {
            minDequeRight--;
        }
        minDeque[minDequeRight++] = nums[right];

        // Check if the current window exceeds the limit
        while ((maxDequeLeft < maxDequeRight) && (minDequeLeft < minDequeRight) &&
               (maxDeque[maxDequeLeft] - minDeque[minDequeLeft] > limit)) {
            // Remove the elements that are out of the current window
            if (maxDeque[maxDequeLeft] == nums[left]) {
                maxDequeLeft++;
            }
            if (minDeque[minDequeLeft] == nums[left]) {
                minDequeLeft++;
            }

            left++;
        }
        retVal = fmax(retVal, right - left + 1);
    }

    return retVal;
}

int main(int argc, char** argv) {
#define MAX_SIZE (int)(1e5)
    struct testCaseType {
        int nums[MAX_SIZE];
        int numsSize;
        int limit;
    } testCase[] = {{{8, 2, 4, 7}, 4, 4}, {{10, 1, 2, 4, 7, 2}, 6, 5}, {{4, 2, 2, 2, 4, 4, 2, 2}, 8, 0}};
    int numberOfTestCase = sizeof(testCase) / sizeof(testCase[0]);
    /* Example
     *  Input: nums = [8,2,4,7], limit = 4
     *  Output: 2
     *
     *  Input: nums = [10,1,2,4,7,2], limit = 5
     *  Output: 4
     *
     *  Input: nums = [4,2,2,2,4,4,2,2], limit = 0
     *  Output: 3
     */

    int answer;
    int i, j;
    for (i = 0; i < numberOfTestCase; ++i) {
        printf("Input: nums = [");
        for (j = 0; j < testCase[i].numsSize; ++j) {
            printf("%s%d", (j == 0) ? "" : ",", testCase[i].nums[j]);
        }
        printf("], limit = %d\n", testCase[i].limit);

        answer = longestSubarray(testCase[i].nums, testCase[i].numsSize, testCase[i].limit);
        printf("Output: %d\n", answer);

        printf("\n");
    }

    return EXIT_SUCCESS;
}
